python/ray/data/dataset_iterator.py

```diff
@@ -0,0 +1,248 @@
+import abc
+from typing import TYPE_CHECKING, Dict, List, Optional, Union, Iterator
+from typing_extensions import Literal
+
+from ray.data.block import DataBatch
+from ray.util.annotations import PublicAPI
+
+if TYPE_CHECKING:
+    import tensorflow as tf
+    import torch
+    from ray.data._internal.torch_iterable_dataset import TorchTensorBatchType
+
+
+@PublicAPI(stability="beta")
+class DatasetIterator(abc.ABC):
+    """An iterator for reading items from a :class:`~Dataset` or
+    :class:`~DatasetPipeline`.
+
+    For Datasets, each iteration call represents a complete read of all items in the
+    Dataset. For DatasetPipelines, each iteration call represents one pass (epoch)
+    over the Dataset. Note that for DatasetPipelines, each pass iterates over
+    the original Dataset, instead of a window (if `.window()` was used).
+
+    If using Ray AIR, each trainer actor should get its own iterator by calling
+    :meth:`session.get_dataset_shard("train")
+    <ray.air.session.get_dataset_shard>`.
+
+    Examples:
+        >>> import ray
+        >>> ds = ray.data.range(5)
+        >>> ds
+        Dataset(num_blocks=5, num_rows=5, schema=<class 'int'>)
+        >>> ds.iterator()
+        DatasetIterator(Dataset(num_blocks=5, num_rows=5, schema=<class 'int'>))
+        >>> ds = ds.repeat(); ds
+        DatasetPipeline(num_windows=inf, num_stages=2)
+        >>> ds.iterator()
+        DatasetIterator(DatasetPipeline(num_windows=inf, num_stages=2))
+
+    Tip: For debugging purposes, use
+    :meth:`~ray.air.util.check_ingest.make_local_dataset_iterator` to create a
+    local `DatasetIterator` from a :class:`~ray.data.Dataset`, a
+    :class:`~ray.data.Preprocessor`, and a :class:`~ray.air.DatasetConfig`.
```

Comment:
Nit: Use Sphinx/RST tip for better rendering in docs.
```suggestion
   .. tip::
       For debugging purposes, use
        :meth:`~ray.air.util.check_ingest.make_local_dataset_iterator` to create a
        local `DatasetIterator` from a :class:`~ray.data.Dataset`, a
        :class:`~ray.data.Preprocessor`, and a :class:`~ray.air.DatasetConfig`.
```